feat: Implement validation to prevent confirming incomplete annotation

teddy-f-47 opened the pull request:
This PR solves two things: (1) Adding the validation mechanism in Summary, Emotions, and Others modes to prevent user from confirming current annotation and going to the next text when they haven't clicked on every slider (or haven't filled the textfields). A warning will appear when they try to do so, but it won't appear in admin account because it may be disturbing during debugging. (2) Some bug fixes related to sliders and checkboxes in Others mode.

What's changed:
 - frontend/pages/projects/_id/affective-annotation/index.vue : Add incomplete annotation validation for Summary, Emotions, Others modes.
 - frontend/i18n/* : Add text warning and its translations for incomplete annotation validation.
 - frontend/components/tasks/affectiveAnnotation/inputs/Slider.vue : Prevent event emit when the value of the checkbox changes during transition to the next text, also add some other minor bug fixes.
 - frontend/components/tasks/affectiveAnnotation/emotions/EmotionsInput.vue : Set default values for sliders to -99. Previously, the default values were set to 0, making it difficult to distinguish a slider that hasn't been clicked at all from a slider that has loaded the value 0 from database. Now, it is possible to distinguish them in order to make the slider's thumb transparent in the case of an unclicked slider. Since the value -99 is only at the beginning when the slider hasn't been clicked at all, that value will not get stored into database.
 - frontend/components/tasks/affectiveAnnotation/others/OthersInput.vue : Similarly, set default values for sliders to -99.

Screenshots:
<img width="960" alt="ss1" src="https://user-images.githubusercontent.com/64476430/200073058-944ddfcc-a264-4124-bb5e-3c1abf884ac5.PNG">
The warning shown when a user tries to confirm without clicking/answering all categories.

<img width="960" alt="ss2" src="https://user-images.githubusercontent.com/64476430/200073083-8a149a75-88a6-49a6-b496-39bbbd088129.PNG">
Initial appearance of sliders that haven't been clicked and haven't loaded any value from db. The slider's thumb is not shown (transparent). This was the initial behavior of the sliders in early development, but got lost due to circumstances that arose from subsequent development that I didn't think about.

artikandri commented on frontend/pages/projects/_id/affective-annotation/index.vue:
```diff
@@ -741,9 +750,30 @@ export default {
+      }
+      if (this.project.isOthersMode) {
+        return (
+          _.keys(this.scaleTypesTextsIds).length === _.keys(this.affectiveScalesValues).length &&
```
I think rather than checking the keys length, its better to check whether the scales values ids are all present in the scaleTypes or not XD    you can use _.intersectionWith or .every for that. If we are using `length` comparison, then if I added a new scale value in the table (because im a curious admin) this check will always be false.

teddy-f-47 replied:
Hello! I agree with your point that checking the length against `scaleTypesTextsIds` is a bad idea because it may contain more scale labels if the admin added their own custom entries. However, I decided to fix it by checking against `affectiveScalesDict`, which is the dict constructed from the static json file inside frontend directory. I think it should work correctly.

I didn't want to go with checking the ids because currently the inputs (sliders) are hard-coded so that if the admin adds additional scale labels, they won't be displayed in annotation page (at least in Emotions and Others modes), and hence checking the ids will cause issues (as there is no way to add responses for these additional scale labels). By checking against the static json file, it should work correctly even if the admin somehow adds additional scale labels, as the most important thing is that the user has provided responses to the dimensions listed in the static json file.

Is it fine? 😅

artikandri commented on frontend/pages/projects/_id/affective-annotation/index.vue:
```diff
@@ -741,9 +750,30 @@ export default {
-      this.updateProgress()
+      const canConfirm = this.isAllAffectiveDataAdded()
+      if (canConfirm || this.isProjectAdmin) {
+        this.cannotConfirmAlert = false
```
It's better to use positive value rather than negative whenever possible XD so its better to use `canConfirm` and discard  `cannotConfirmAlert` and use `canConfirm` in template, what do you think?

artikandri replied:
or better, you can show the alert only when `canConfirm` is false and user `hasClickedReviewButton` XD

teddy-f-47 replied:
noted, thank you for pointing these things out XD I will clean the code following your suggestions 😅

teddy-f-47 replied:
I have refactored the code related to this comment :)

artikandri reviewed (commented):
Hello, thank you for the good work XD I have some comments, can you check it  XD

haanh764 reviewed (commented):
Hello, thank you for your good work XD I've tested and it works well in general.
I have only 2 very minor comment XD
- In summary mode, we require user to enter at least 2 words for each textfield. So if I enter 1 word only, it will be red and warn me that I need to enter at least 2 words, like this 
![image](https://user-images.githubusercontent.com/17596118/200283664-f52a7cb5-98ce-4c87-addc-774baf531448.png)
but if I enter 2 words but in 1 line (I mean at once and separate them by ',') - like the picture below. There's no red text for warning, everything has been saved in database but I can't confirm it :sweat_smile: So it's rather a question that this situation is acceptable or not(?) If it's acceptable to enter 2 or more words at once with "," as separator then we should make it's oke to click on confirm button. If it's not acceptable, then we should add a warning for it XD Wdyt?
![image](https://user-images.githubusercontent.com/17596118/200284235-e31183aa-b344-402d-9c02-a4f7847d3c18.png)
- 2nd point is just a question also XD I think it'll be good if the user can close the warning. I mean when I try to confirm without finishing, the warning shows up, then it's good that I can be able to close it (by 'x' button for eg), because seeing it all the time is a bit scary and distraction for user experience XD How do you think :smile:

teddy-f-47 commented:
>   but if I enter 2 words but in 1 line (I mean at once and separate them by ',') - like the picture below. There's no red text for warning, everything has been saved in database but I can't confirm it 😅 So it's rather a question that this situation is acceptable or not(?) If it's acceptable to enter 2 or more words at once with "," as separator then we should make it's oke to click on confirm button. If it's not acceptable, then we should add a warning for it XD Wdyt?

1) I think entering 2 or more words at once with "," is not acceptable XD because that way I'm not sure if in the db the words are saved in one record as a long string, or as individual words. Apparently I missed excluding "," from the allowed character list, I will have a look at that. XD

> * 2nd point is just a question also XD I think it'll be good if the user can close the warning. I mean when I try to confirm without finishing, the warning shows up, then it's good that I can be able to close it (by 'x' button for eg), because seeing it all the time is a bit scary and distraction for user experience XD How do you think 😄

2) I see, you have a point 😅 I will adjust it by adding a close button XD

haanh764 commented:
> > but if I enter 2 words but in 1 line (I mean at once and separate them by ',') - like the picture below. There's no red text for warning, everything has been saved in database but I can't confirm it sweat_smile So it's rather a question that this situation is acceptable or not(?) If it's acceptable to enter 2 or more words at once with "," as separator then we should make it's oke to click on confirm button. If it's not acceptable, then we should add a warning for it XD Wdyt?
> 
> 1. I think entering 2 or more words at once with "," is not acceptable XD because that way I'm not sure if in the db the words are saved in one record as a long string, or as individual words. Apparently I missed excluding "," from the allowed character list, I will have a look at that. XD
> 
Yeah, from database it's saved as a long string like this `fun, admire` fyi :smile:  but I dont think it's a good idea also, separate the words are better XD

teddy-f-47 commented:
> Yeah, from database it's saved as a long string like this `fun, admire` fyi 😄 but I dont think it's a good idea also, separate the words are better XD

Hello, I have addressed these points. So, for (1) I have removed comma "," from the allowed characters so now it should work correctly to make the user input the keyword one by one. In addition, I have also added a hint about how to input the keywords and the allowed chars: (note: I allow space and hyphen because I think a keyword may consist of two or three words, but we can also remove these chars if we want keywords to be single words)
<img width="960" alt="ss1" src="https://user-images.githubusercontent.com/64476430/200319191-7505170c-5048-4f5a-98c8-a7566e46bfd7.PNG">

And for (2), I have added the close button to the alert message 😅 
<img width="960" alt="ss2" src="https://user-images.githubusercontent.com/64476430/200319393-943d0e05-1f01-471e-9a7f-361101383b1b.PNG">

haanh764 reviewed (approved):
Hello, I've checked again and everything works fine for me XD I also agree that the blank space is ok because sometimes it's not a single word. So approved for merging XD

artikandri reviewed (approved):
Hello, I have tested the code and everything works well so I can give the approval, good job! However, it should be noted that the `keys.length` is not a good solution, especially after today's  request  :"   but since there is no way to handle it for now, it can be merged and fixed later. Thank you!

teddy-f-47 commented:
> Hello, I have tested the code and everything works well so I can give the approval, good job! However, it should be noted that the `keys.length` is not a good solution, especially after today's request :" but since there is no way to handle it for now, it can be merged and fixed later. Thank you!

You mean the request about enabling multiple modes in one project, right? 😅  I can imagine the possible issue...., I think I can change it to checking each label in the static json file. But I will merge the PR for now, and do the adjustment in the future when we start working on the request. Thank you for the feedback!
